### PR TITLE
Advance resource management 

### PR DESCRIPTION
**Pull request by qzhuyan**
Tend to decopule the lifetime of msquic handle and nif resource context. 
The resource in msquic could be released earlier before the deallocation of erlang NIF resource with some cost of extra refcnts overheads.

Benefits:
1. For correctness, do not rely on the internal ref countings in MsQuic, take it as black box. 
2. Further reduce memory footprints at server side, adapt to more dynamic environments. 
3. simplify the logic of resource management at both msquic side and erlang nif side. 
4. make it more flexible to operate the stack, support more dynamic configuration. 

TODOs:
- [ ] list Breaking changes:
- [x] update API document
- [x] update internal dev doc
- [x] check coverage

targeting ver: 0.2

### CODE REVIEW

File: c_src/quicer_listener.c [hunk 1]
```diff
@@ -339,50 +324,53 @@ listen2(ErlNifEnv *env, __unused_parm__ int argc, const ERL_NIF_TERM argv[])
...
   if (l_ctx->r_ctx)
     {
-      // quic_registration is set
-      enif_keep_resource(l_ctx->r_ctx);
+      // quic_registration is set,
+      // none-global registration.
+      CXPLAT_DBG_ASSERT(l_ctx->r_ctx != &G_r_ctx);
       if (!get_reg_handle(l_ctx->r_ctx))
         {
+          // avoid destroy_l_ctx to put again it.
```
Comment (qzhuyan) [inline c_src/quicer_listener.c:332]: remove comment

File: c_src/quicer_listener.c [hunk 2]
```diff
@@ -391,7 +379,7 @@ listen2(ErlNifEnv *env, __unused_parm__ int argc, const ERL_NIF_TERM argv[])
           env, l_ctx, &l_ctx->listenerPid, &l_ctx->owner_mon))
     {
       ret = ERROR_TUPLE_2(ATOM_BAD_MON);
-      goto exit;
+      goto exit_no_reg;
```
Comment (qzhuyan) [inline c_src/quicer_listener.c:382]:
```suggestion
      put_config_handle(l_ctx->config_ctx);
      l_ctx->config_ctx = NULL;
      goto exit_no_reg;
```

File: c_src/quicer_listener.c [hunk 3]
```diff
@@ -339,50 +324,53 @@ listen2(ErlNifEnv *env, __unused_parm__ int argc, const ERL_NIF_TERM argv[])
...
   ret = ServerLoadConfiguration(env,
                                 &options,
                                 Registration,
-                                &l_ctx->config_resource->Configuration,
+                                &l_ctx->config_ctx->Configuration,
                                 &CredConfig);
   if (!IS_SAME_TERM(ATOM_OK, ret))
     {
       // @TODO unsure 3 elem tuple is the best way to return error
+      put_config_handle(l_ctx->config_ctx);
```
Comment (qzhuyan) [inline c_src/quicer_listener.c:371]:
```suggestion
      put_config_handle(l_ctx->config_ctx);
      l_ctx->config_ctx = NULL;
```

File: c_src/quicer_listener.c [hunk 4]
```diff
@@ -458,12 +435,17 @@ listen2(ErlNifEnv *env, __unused_parm__ int argc, const ERL_NIF_TERM argv[])
   free_certificate(&CredConfig);
   return OK_TUPLE_2(listenHandle);
 
+exit_no_reg:
```
Comment (qzhuyan) [inline c_src/quicer_listener.c:438]: I think we don't need this part, just call the put_listener_handle(l_ctx);

File: c_src/quicer_listener.c [hunk 5]
```diff
@@ -473,36 +455,27 @@ close_listener1(ErlNifEnv *env,
...
 
   enif_mutex_unlock(l_ctx->lock);
 
-  MsQuic->ListenerClose(l);
-  if (is_destroy)
-    {
-      destroy_l_ctx(l_ctx);
-    }
+  put_listener_handle(l_ctx);
+  ret = ATOM_CLOSED;
```
Comment (qzhuyan) [inline c_src/quicer_listener.c:478]: update spec for this change.

File: c_src/quicer_ctx.c [hunk 6]
```diff
@@ -202,26 +182,14 @@ destroy_c_ctx(QuicerConnCTX *c_ctx)
...
-  else
-    {
-      r_ctx = G_r_ctx;
-    }
-
-  enif_mutex_lock(r_ctx->lock);
-  CxPlatListEntryRemove(&c_ctx->RegistrationLink);
-  enif_mutex_unlock(r_ctx->lock);
 
   if (c_ctx->is_monitored)
```
[the remark this answers is not in the document]
Comment (qzhuyan) [reply c_src/quicer_ctx.c:186]: still need it, we don't want dead proc trigger the unmanaged closing.

File: c_src/quicer_ctx.c [hunk 7]
```diff
@@ -400,13 +412,47 @@ get_listener_handle(QuicerListenerCTX *l_ctx)
...
+}
+
+inline void
+put_config_handle(QuicerConfigCTX *config_ctx)
+{
+  if (CxPlatRefDecrement(&config_ctx->ref_count))
+    {
+      MsQuic->ConfigurationClose(config_ctx->Configuration);
+      config_ctx->Configuration = NULL;
+      enif_release_resource(config_ctx);
```
[the remark this answers is not in the document]
Comment (qzhuyan) [reply c_src/quicer_ctx.c:445]: destory_config_ctx is removed now.

**Review by zmstone — APPROVED** [review]

**Pull request merged**